pkg/controller/beat/controller.go

```diff
@@ -0,0 +1,278 @@
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/annotation"
+	commonbeat "github.com/elastic/cloud-on-k8s/pkg/controller/common/beat"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/beat/filebeat"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/beat/otherbeat"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/events"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/operator"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/reconciler"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/tracing"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/watches"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
+)
+
+const (
+	controllerName = "beat-controller"
+)
+
+var log = logf.Log.WithName(controllerName)
+
+// Add creates a new Beat Controller and adds it to the Manager with default RBAC. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager, params operator.Parameters) error {
+	r := newReconciler(mgr, params)
+	c, err := common.NewController(mgr, controllerName, r, params)
+	if err != nil {
+		return err
+	}
+	return addWatches(c, r)
+}
+
+// newReconciler returns a new reconcile.Reconciler.
+func newReconciler(mgr manager.Manager, params operator.Parameters) *ReconcileBeat {
+	client := k8s.WrapClient(mgr.GetClient())
+	return &ReconcileBeat{
+		Client:         client,
+		recorder:       mgr.GetEventRecorderFor(controllerName),
+		dynamicWatches: watches.NewDynamicWatches(),
+		Parameters:     params,
+	}
+}
+
+// addWatches registers the required watches.
+func addWatches(c controller.Controller, r *ReconcileBeat) error {
+	// Watch for changes to Beat
+	if err := c.Watch(&source.Kind{Type: &beatv1beta1.Beat{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &appsv1.DaemonSet{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &beatv1beta1.Beat{},
+	}); err != nil {
+		return err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &appsv1.Deployment{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &beatv1beta1.Beat{},
+	}); err != nil {
+		return err
+	}
+
+	// Watch secrets
+	if err := c.Watch(&source.Kind{Type: &corev1.Secret{}}, r.dynamicWatches.Secrets); err != nil {
+		return err
+	}
+	if err := r.dynamicWatches.Secrets.AddHandler(&watches.OwnerWatch{
+		EnqueueRequestForOwner: handler.EnqueueRequestForOwner{
+			IsController: true,
+			OwnerType:    &beatv1beta1.Beat{},
+		},
+	}); err != nil {
+		return err
+	}
+
+	if commonbeat.ShouldSetupAutodiscoverRBAC() {
+		if err := c.Watch(&source.Kind{Type: &corev1.ServiceAccount{}}, &handler.EnqueueRequestForOwner{
+			IsController: true,
+			OwnerType:    &beatv1beta1.Beat{},
+		}); err != nil {
+			return err
+		}
+
+		if err := c.Watch(&source.Kind{Type: &rbacv1.ClusterRoleBinding{}}, &handler.EnqueueRequestForObject{}); err != nil {
```

Comment:
This will trigger a reconcile loop for each `ClusterRoleBinding`, I understand that the idea is to ensure that `ServiceAccounts` and `ClusterRoleBinding` are reconciled but it is a little bit "surprising" to see all those messages about some `beat_name` that don't exist during the startup:

```
2020-06-02T15:01:49.397+0200    INFO    beat-controller Starting reconciliation run     {"service.version": "1.1.0-SNAPSHOT-mmo", "iteration": 1, "namespace": "", "beat_name": "system:openshift:controller:unidling-controller"}
2020-06-02T15:01:49.397+0200    INFO    beat-controller Starting reconciliation run     {"service.version": "1.1.0-SNAPSHOT-mmo", "iteration": 2, "namespace": "", "beat_name": "csi-snapshot-controller-operator-role"}
2020-06-02T15:03:58.111+0200    INFO    beat-controller Ending reconciliation run       {"service.version": "1.1.0-SNAPSHOT-mmo", "iteration": 2, "namespace": "", "beat_name": "csi-snapshot-controller-operator-role", "took": "2m8.710463019s"}
2020-06-02T15:03:58.111+0200    INFO    beat-controller Ending reconciliation run       {"service.version": "1.1.0-SNAPSHOT-mmo", "iteration": 1, "namespace": "", "beat_name": "system:openshift:controller:unidling-controller", "took": "2m8.710491152s"}
2020-06-02T15:03:58.111+0200    INFO    beat-controller Starting reconciliation run     {"service.version": "1.1.0-SNAPSHOT-mmo", "iteration": 3, "namespace": "", "beat_name": "system:openshift:controller:build-config-change-controller"}
2020-06-02T15:03:58.112+0200    INFO    beat-controller Starting reconciliation run     {"service.version": "1.1.0-SNAPSHOT-mmo", "iteration": 4, "namespace": "", "beat_name": "cluster-autoscaler"}
2020-06-02T15:03:59.292+0200    INFO    beat-controller Ending reconciliation run       {"service.version": "1.1.0-SNAPSHOT-mmo", "iteration": 3, "namespace": "", "beat_name": "system:openshift:controller:build-config-change-controller", "took": "1.181247897s"}
2020-06-02T15:03:59.292+0200    INFO    beat-controller Starting reconciliation run     {"service.version": "1.1.0-SNAPSHOT-mmo", "iteration": 5, "namespace": "", "beat_name": "system:openshift:operator:kube-apiserver-operator"}
2020-06-02T15:03:59.836+0200    INFO    beat-controller Ending reconciliation run       {"service.version": "1.1.0-SNAPSHOT-mmo", "iteration": 4, "namespace": "", "beat_name": "cluster-autoscaler", "took": "1.72457487s"}
2020-06-02T15:04:00.380+0200    INFO    beat-controller Starting reconciliation run     {"service.version": "1.1.0-SNAPSHOT-mmo", "iteration": 6, "namespace": "", "beat_name": "system:controller:disruption-controller"}
2020-06-02T15:04:00.380+0200    INFO    beat-controller Starting reconciliation run     {"service.version": "1.1.0-SNAPSHOT-mmo", "iteration": 7, "namespace": "", "beat_name": "kube-apiserver"}
2020-06-02T15:04:00.380+0200    INFO    beat-controller Ending reconciliation run       {"service.version": "1.1.0-SNAPSHOT-mmo", "iteration": 5, "namespace": "", "beat_name": "system:openshift:operator:kube-apiserver-operator", "took": "1.087855122s"}
2020-06-02T15:04:00.827+0200    INFO    beat-controller Ending reconciliation run       {"service.version": "1.1.0-SNAPSHOT-mmo", "iteration": 7, "namespace": "", "beat_name": "kube-apiserver", "took": "447.076399ms"}
2020-06-02T15:04:00.827+0200    INFO    beat-controller Starting reconciliation run     {"service.version": "1.1.0-SNAPSHOT-mmo", "iteration": 8, "namespace": "", "beat_name": "alertmanager-main"}
2020-06-02T15:04:01.301+0200    INFO    beat-controller Ending reconciliation run       {"service.version": "1.1.0-SNAPSHOT-mmo", "iteration": 8, "namespace": "", "beat_name": "alertmanager-main", "took": "473.845307ms"}
2020-06-02T15:04:01.301+0200    INFO    beat-controller Starting reconciliation run     {"service.version": "1.1.0-SNAPSHOT-mmo", "iteration": 9, "namespace": "", "beat_name": "default-account-cluster-network-operator"}
2020-06-02T15:04:00.827+0200    INFO    beat-controller Ending reconciliation run       {"service.version": "1.1.0-SNAPSHOT-mmo", "iteration": 6, "namespace": "", "beat_name": "system:controller:disruption-controller", "took": "447.267043ms"}
2020-06-02T15:04:01.739+0200    INFO    beat-controller Starting reconciliation run     {"service.version": "1.1.0-SNAPSHOT-mmo", "iteration": 10, "namespace": "", "beat_name": "system:controller:generic-garbage-collector"}
2020-06-02T15:04:09.500+0200    INFO    beat-controller Ending reconciliation run       {"service.version": "1.1.0-SNAPSHOT-mmo", "iteration": 9, "namespace": "", "beat_name": "default-account-cluster-network-operator", "took": "8.198567717s"}
2020-06-02T15:04:09.500+0200    INFO    beat-controller Starting reconciliation run     {"service.version": "1.1.0-SNAPSHOT-mmo", "iteration": 11, "namespace": "", "beat_name": "system:scope-impersonation"}
```

I'm wondering if it's worth it to introduce a dedicated controller to reconcile these resources.

Reply:
We could also filter on labels here so we only reconcile on ECK-managed `ClusterRoleBinding`?

Reply:
I did the filtering that @sebgl suggested.